indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/InputSourceSampler.java

```diff
@@ -0,0 +1,227 @@
+import org.apache.druid.segment.indexing.DataSchema;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class InputSourceSampler
+{
+  private static final String SAMPLER_DATA_SOURCE = "sampler";
+
+  private static final DataSchema DEFAULT_DATA_SCHEMA = new DataSchema(
+      SAMPLER_DATA_SOURCE,
+      new TimestampSpec(null, null, null),
+      new DimensionsSpec(null),
+      null,
+      null,
+      null
+  );
+
+  // We want to be able to sort the list of processed results back into the same order that we read them from the
+  // firehose so that the rows in the data loader are not always changing. To do this, we add a temporary column to the
+  // InputRow (in SamplerInputRow) and tag each row with a sortKey. We use an aggregator so that it will not affect
+  // rollup, and we use a longMin aggregator so that as rows get rolled up, the earlier rows stay stable and later
+  // rows may get rolled into these rows. After getting the results back from the IncrementalIndex, we sort by this
+  // column and then exclude it from the response.
+  private static final AggregatorFactory INTERNAL_ORDERING_AGGREGATOR = new LongMinAggregatorFactory(
+      SamplerInputRow.SAMPLER_ORDERING_COLUMN,
+      SamplerInputRow.SAMPLER_ORDERING_COLUMN
+  );
+
+  public SamplerResponse sample(
+      final InputSource inputSource,
+      // inputFormat can be null only if inputSource.needsFormat() = false or parser is specified.
+      @Nullable final InputFormat inputFormat,
+      @Nullable final DataSchema dataSchema,
+      @Nullable final SamplerConfig samplerConfig
+  )
+  {
+    Preconditions.checkNotNull(inputSource, "inputSource required");
+    if (inputSource.needsFormat()) {
+      Preconditions.checkNotNull(inputFormat, "inputFormat required");
+    }
+    final DataSchema nonNullDataSchema = dataSchema == null
+                                         ? DEFAULT_DATA_SCHEMA
+                                         : dataSchema;
+    final SamplerConfig nonNullSamplerConfig = samplerConfig == null
+                                               ? SamplerConfig.empty()
+                                               : samplerConfig;
+
+    final Closer closer = Closer.create();
+    final File tempDir = Files.createTempDir();
+    closer.register(() -> FileUtils.deleteDirectory(tempDir));
+
+    final InputSourceReader reader = buildReader(
+        nonNullSamplerConfig,
+        nonNullDataSchema,
+        inputSource,
+        inputFormat,
+        tempDir
+    );
+    try (final CloseableIterator<InputRowListPlusRawValues> iterator = reader.sample();
+         final IncrementalIndex<Aggregator> index = buildIncrementalIndex(nonNullSamplerConfig, nonNullDataSchema);
+         final Closer closer1 = closer) {
```

Comment:
This with the closer seems kind of sad,  but no great solutions since would be sort of funny to register the tmp dir inside the try after already using it.

Reply:
Yeah, but we can tidy up later once we move to Java 11 (https://docs.oracle.com/javase/specs/jls/se9/html/jls-14.html#jls-14.20.3).